File: translate.mjs
import { readFileSync, writeFileSync } from 'fs';
import fetch from 'node-fetch';

// Regular expression to match YAML front matter (enclosed by ---)
const yamlFrontMatterRegex = /^---\n([\s\S]*?)\n---/;

async function translateText(text, targetLanguage, sourceLanguage = 'en', apiKey) {
  const url = `https://translation.googleapis.com/language/translate/v2?key=${apiKey}`;
  const body = JSON.stringify({ q: text, target: targetLanguage, source: sourceLanguage, format: 'text' });

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: body
    });
    const data = await response.json();
    if (response.ok) {
      return data.data.translations[0].translatedText;
    } else {
      throw new Error(data.error.message);
    }
  } catch (error) {
    console.error('Error translating text:', error);
    return null;
  }
}

const codeBlockRegex = /(```[\s\S]*?```|<script[\s\S]*?>[\s\S]*?<\/script>|<code[\s\S]*?>[\s\S]*?<\/code>)/g;

/* 🆕 [ADDED FUNCTION] Add lang parameter to all URLs in the text */
function addLangParamToUrls(text, langCode) {
  // Match typical URLs (http, https)
  const urlRegex = /(https?:\/\/[^\s)'"<>]+)/g;
  return text.replace(urlRegex, (url) => {
    try {
      const urlObj = new URL(url);
      // Avoid duplicates if lang already exists
      if (!urlObj.searchParams.has('lang')) {
        urlObj.searchParams.append('lang', langCode);
      }
      return urlObj.toString();
    } catch {
      return url; // leave invalid URLs untouched
    }
  });
}
/* 🆕 [END OF ADDED FUNCTION] */

async function translateMarkdownFile(inputFile, outputFile, targetLanguage, apiKey) {
  try {
    // Read the file content
    const content = readFileSync(inputFile, 'utf8');
    if (typeof content !== 'string') {
      throw new Error('File content is not a string');
    }

    // Extract YAML front matter (if any)
    const match = content.match(yamlFrontMatterRegex);
    let yamlFrontMatter = '';
    let modifiedContent = content;

    if (match) {
      yamlFrontMatter = match[0]; // Capture the YAML front matter
      modifiedContent = content.replace(yamlFrontMatterRegex, ''); // Remove it from the content for translation
    }

    // Detect and extract code blocks
    const codeBlocks = [];
    const modifiedContentWithoutCodeBlocks = modifiedContent.replace(codeBlockRegex, (match) => {
      codeBlocks.push(match);
      return `<!--CODEBLOCK_${codeBlocks.length - 1}-->`; // Placeholder for code blocks
    });

    // Split the content into lines and analyze
    const lines = modifiedContentWithoutCodeBlocks.split('\n').map(line => {
      const matchIndentation = line.match(/^(\s*)/); // Capture leading whitespace
      const leadingWhitespace = matchIndentation ? matchIndentation[0] : '';

      if (line.startsWith('#') || line.startsWith('##')) {
        return { original: line, translated: line, leadingWhitespace }; // Keep headers as is
      } else if (line.trim() === '' || line.includes('<!--CODEBLOCK_')) {
        return { original: line, translated: line, leadingWhitespace }; // Skip placeholders or blank lines
      } else {
        return { original: line, translated: null, leadingWhitespace }; // Mark for translation
      }
    });

    // Extract text to translate
    const textToTranslate = lines
      .filter(line => line.translated === null)
      .map(line => line.original.trim()) // Remove leading/trailing whitespace for translation
      .join('\n');

    // Translate the text
    const translatedText = await translateText(textToTranslate, targetLanguage, 'en', apiKey);

    if (translatedText) {
      // Split the translated text into lines
      const translatedLines = translatedText.split('\n');
      let translatedLineIndex = 0;

      // Rebuild the final content with preserved indentation and placeholders
      const finalContent = lines.map(line => {
        if (line.translated === null) {
          const translated = translatedLines[translatedLineIndex++];
          return `${line.leadingWhitespace}${translated}`; // Reapply leading whitespace
        } else {
          return line.original; // Keep headers, blank lines, and placeholders intact
        }
      }).join('\n');

      // Reinsert code blocks into placeholders
      const finalContentWithCodeBlocks = finalContent.replace(/<!--CODEBLOCK_(\d+)-->/g, (_, index) => codeBlocks[index]);

     const finalContentWithLangUrls = finalContentWithCodeBlocks.replace(
  /https:\/\/mnpt-local-dev\.o18-test\.com\/new-server\/year\/2025\/oct\/script\.php\?lang=en/g,
  `https://mnpt-local-dev.o18-test.com/new-server/year/2025/oct/script.php?lang=${targetLanguage}`
);
      /* 🆕 [ADDED HERE] Add language parameter to URLs before writing the file */
      // const finalContentWithLangUrls = addLangParamToUrls(finalContentWithCodeBlocks, targetLanguage);
      /* 🆕 [END OF ADDITION] */

      // Combine YAML front matter with the translated content
      const finalContentWithYaml = yamlFrontMatter + '\n' + finalContentWithLangUrls;

      // Write to the output file
      writeFileSync(outputFile, finalContentWithYaml, 'utf8');
      console.log('Translation complete. File saved as:', outputFile);
    } else {
      console.error('Translation failed. No changes were written.');
    }
  } catch (error) {
    console.error('Error processing the markdown file:', error);
  }
}

// Read command-line arguments
const inputFile = process.argv[2];
const outputFile = process.argv[3];
const targetLanguage = process.argv[4];
const apiKey = process.argv[5];

if (!inputFile || !outputFile || !targetLanguage || !apiKey) {
  console.error('Usage: node translate.mjs <inputFile> <outputFile> <targetLanguage> <apiKey>');
  process.exit(1);
}

translateMarkdownFile(inputFile, outputFile, targetLanguage, apiKey);

File: translate.py
import os
from google.cloud import translate_v2 as translate

def translate_text(target: str, text: str, api_key: str):
    """Translates text into the target language.

    Target must be an ISO 639-1 language code.
    See https://g.co/cloud/translate/v2/translate-reference#supported_languages

    Args:
        target (str): ISO 639-1 language code for translation.
        text (str): Text to be translated.
        api_key (str): Your Google Cloud API key.

    Returns:
        str: The translated text.
    """
    translate_client = translate.Client(credentials=translate.Credentials(api_key))

    if isinstance(text, bytes):
        text = text.decode("utf-8")

    result = translate_client.translate(text, target_language=target)
    print(f"Translating text: {text[:100]}... to language: {target}")
    print(f"Translation result: {result['translatedText']}")

    return result['translatedText']

def translate_md_files_in_directory(target_language: str, api_key: str):
    """Translates all Markdown (.md) files in the 'en' directory to the target language."""
    source_dir = "en"
    target_dir = target_language

    if not os.path.exists(source_dir):
        print(f"Source directory '{source_dir}' does not exist.")
        return

    os.makedirs(target_dir, exist_ok=True)

    for root, _, files in os.walk(source_dir):
        for file in files:
            if file.endswith(".md"):
                source_file = os.path.join(root, file)
                relative_path = os.path.relpath(source_file, source_dir)
                target_file = os.path.join(target_dir, relative_path)

                os.makedirs(os.path.dirname(target_file), exist_ok=True)

                with open(source_file, "r", encoding="utf-8") as f:
                    content = f.read()

                translated_content = translate_text(target_language, content, api_key)

                with open(target_file, "w", encoding="utf-8") as f:
                    f.write(translated_content)

                print(f"Translated {source_file} -> {target_file}")

if __name__ == "__main__":
    api_key = input("Enter your Google Cloud API key: ")
    target_language = input("Enter the target language code (e.g., 'es' for Spanish): ").strip()
    translate_md_files_in_directory(target_language, api_key)
